File: day12/Main.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.lang.reflect.Array;
import java.text.NumberFormat;
import java.text.ParseException;
import java.util.*;

public class Main {

    public static int stoi(String s, int beginIndex) throws ParseException {
        NumberFormat nf = NumberFormat.getInstance();
        return nf.parse((s.substring(beginIndex)).strip()).intValue();
    }

    public static long arrangements_dp(long[][][] dp, char[] spring, int[] groupSizes, int[] currSizes, int springIdx, int groupIdx) {
        int si = springIdx;
        int gi = groupIdx;
        int curr = currSizes[groupIdx];

        if (springIdx == spring.length) {
            if (groupIdx == groupSizes.length ||
                    (groupIdx + 1 == groupSizes.length && groupSizes[groupIdx] == currSizes[groupIdx])) {
                return dp[si][gi][curr] = 1;
            }
            return dp[si][gi][curr] = 0;
        }

        if (dp[si][gi][curr] != -1) {
            return dp[si][gi][curr];
        }

        if (groupIdx == groupSizes.length) {
            if (spring[springIdx] == '#') {
                return dp[si][gi][curr] = 0;
            }

            return dp[si][gi][curr] = arrangements_dp(dp, spring, groupSizes, currSizes, springIdx + 1, groupIdx);
        }

        if (spring[springIdx] == '.') {
            if (currSizes[groupIdx] != 0) {
                if (currSizes[groupIdx] != groupSizes[groupIdx]) {
                    return dp[si][gi][curr] = 0;
                }
                groupIdx++;
            }

            return dp[si][gi][curr] = arrangements_dp(dp, spring, groupSizes, currSizes, springIdx + 1, groupIdx);
        }

        if (spring[springIdx] == '#') {
            int[] newCurrSizes = currSizes.clone();
            newCurrSizes[groupIdx]++;
            if (groupSizes[groupIdx] < newCurrSizes[groupIdx]) {
                return dp[si][gi][curr] = 0;
            }

            return dp[si][gi][curr] = arrangements_dp(dp, spring, groupSizes, newCurrSizes, springIdx + 1, groupIdx);
        }

        if (spring[springIdx] == '?') {
            if (currSizes[groupIdx] != 0) {
                if (currSizes[groupIdx] < groupSizes[groupIdx]) {
                    int[] newCurrSizes = currSizes.clone();
                    newCurrSizes[groupIdx]++;
                    return dp[si][gi][curr] = arrangements_dp(dp, spring, groupSizes, newCurrSizes, springIdx + 1, groupIdx);
                }
                if (currSizes[groupIdx] == groupSizes[groupIdx]) {
                    groupIdx++;
                }
                return dp[si][gi][curr] = arrangements_dp(dp, spring, groupSizes, currSizes, springIdx + 1, groupIdx);
            }

            long part = arrangements_dp(dp, spring, groupSizes, currSizes, springIdx + 1, groupIdx);
            int[] newCurrSizes = currSizes.clone();
            newCurrSizes[groupIdx]++;
            return dp[si][gi][curr] = part + arrangements_dp(dp, spring, groupSizes, newCurrSizes, springIdx + 1, groupIdx);
        }

        return 0;
    }

    public static void part1(BufferedReader br) throws Exception {
        String line;
        ArrayList<char[]> springs = new ArrayList<>();
        ArrayList<int[]> groupSizes = new ArrayList<>();

        int repeat = 5;
        while ((line = br.readLine()) != null) {
            String[] tmp = line.split(" ");
            StringBuilder stringBuilder = new StringBuilder(tmp[0]);
            for (int i = 1; i < repeat; i++) {
                stringBuilder.append("?").append(tmp[0]);
            }
            springs.add(stringBuilder.toString().toCharArray());

            String[] nums = tmp[1].split(",");
            ArrayList<Integer> sizes = new ArrayList<>();
            for (String num : nums) {
                sizes.add(stoi(num, 0));
            }
            ArrayList<Integer> groupSize = new ArrayList<>();
            for (int i = 0; i < repeat; i++) {
                groupSize.addAll(sizes);
            }

            groupSizes.add(groupSize.stream().mapToInt(Integer::intValue).toArray());
        }

        long ans = 0;
        for (int i = 0; i < springs.size(); i++) {
            int[] currSizes = new int[groupSizes.get(i).length + 1];
            long[][][] dp = new long[springs.get(i).length + 1][groupSizes.get(i).length + 1][springs.get(i).length + 1];
            for (int j = 0; j < dp.length; j++) {
                for (int k = 0; k < dp[0].length; k++) {
                    Arrays.fill(dp[j][k], -1);
                }
            }

            long ways = arrangements_dp(dp, springs.get(i), groupSizes.get(i), currSizes, 0, 0);
            ans += ways;
        }
        System.out.println(ans);

    }


    public static void part2(BufferedReader br) throws Exception {

    }


    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader("jaoc23/src/input"));
        part1(br);
//        part2(br);

    }
}

File: day12/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <deque>
#include <numeric>
#include "utils_aoc.h"

using namespace std;

vector<vector<vector<long long>>> dp;

long long
arrangements_dp(const string &spring, const vector<int> &group_sizes, vector<int> curr_sizes, int spring_i,
                int group_i) {
    int si = spring_i;
    int gi = group_i;
    int curr = curr_sizes[group_i];

    if (spring_i == spring.size()) {
        if (group_i == group_sizes.size()) {
            return 1;
        }
        if (group_i + 1 == group_sizes.size() && group_sizes[group_i] == curr_sizes[group_i]) {
            return 1;
        }
        return 0;
    }

    if (dp[si][gi][curr] != -1) {
        return dp[si][gi][curr];
    }

    if (group_i == group_sizes.size()) {
        if (spring[spring_i] == '#') {
            return 0;
        }

        return dp[si][gi][curr] = arrangements_dp(spring, group_sizes, curr_sizes, spring_i + 1, group_i);
    }

    if (spring[spring_i] == '.') {
        if (curr_sizes[group_i] != 0) {
            if (curr_sizes[group_i] != group_sizes[group_i]) {
                return 0;
            }
            group_i++;
        }

        return dp[si][gi][curr] = arrangements_dp(spring, group_sizes, curr_sizes, spring_i + 1, group_i);
    }

    if (spring[spring_i] == '#') {
        curr_sizes[group_i]++;
        if (group_sizes[group_i] < curr_sizes[group_i]) {
            return 0;
        }

        return dp[si][gi][curr] = arrangements_dp(spring, group_sizes, curr_sizes, spring_i + 1, group_i);
    }

    if (spring[spring_i] == '?') {
        if (curr_sizes[group_i] != 0) {
            if (curr_sizes[group_i] == group_sizes[group_i]) {
                group_i++;
            } else if (curr_sizes[group_i] < group_sizes[group_i]) {
                curr_sizes[group_i]++;
            }
            return dp[si][gi][curr] = arrangements_dp(spring, group_sizes, curr_sizes, spring_i + 1, group_i);
        }

        long long part = arrangements_dp(spring, group_sizes, curr_sizes, spring_i + 1, group_i);
        curr_sizes[group_i]++;
        return dp[si][gi][curr] = part + arrangements_dp(spring, group_sizes, curr_sizes, spring_i + 1, group_i);
    }

    return 0;
}


long long
arrangements(const string &spring, const vector<int> &group_sizes, vector<int> curr_sizes, int spring_i, int group_i) {
    if (spring_i == spring.size()) {
        if (group_i == group_sizes.size()) {
            return 1;
        }
        if (group_i + 1 == group_sizes.size() && group_sizes[group_i] == curr_sizes[group_i]) {
            return 1;
        }
        return 0;
    }

    if (group_i == group_sizes.size()) {
        if (spring[spring_i] == '#') {
            return 0;
        }

        return arrangements(spring, group_sizes, curr_sizes, spring_i + 1, group_i);
    }

    if (spring[spring_i] == '.') {
        if (curr_sizes[group_i] != 0) {
            if (curr_sizes[group_i] != group_sizes[group_i]) {
                return 0;
            }
            group_i++;
        }

        return arrangements(spring, group_sizes, curr_sizes, spring_i + 1, group_i);
    }

    if (spring[spring_i] == '#') {
        curr_sizes[group_i]++;
        if (group_sizes[group_i] < curr_sizes[group_i]) {
            return 0;
        }

        return arrangements(spring, group_sizes, curr_sizes, spring_i + 1, group_i);
    }

    if (spring[spring_i] == '?') {
        if (curr_sizes[group_i] != 0) {
            if (curr_sizes[group_i] == group_sizes[group_i]) {
                group_i++;
            } else if (curr_sizes[group_i] < group_sizes[group_i]) {
                curr_sizes[group_i]++;
            }
            return arrangements(spring, group_sizes, curr_sizes, spring_i + 1, group_i);
        }

        long long part = arrangements(spring, group_sizes, curr_sizes, spring_i + 1, group_i);
        curr_sizes[group_i]++;
        return part + arrangements(spring, group_sizes, curr_sizes, spring_i + 1, group_i);
    }

    cout << "err - function end" << endl;
    return 0;
}

void part1(ifstream &input) {
    string line;
    vector<string> springs;
    vector<vector<int>> group_sizes;
    while (getline(input, line)) {
        auto sp = filter(split(line, " "));
        springs.push_back(sp[0]);
        vector<int> tmp = stoi(filter(split(sp[1], ",")));
        group_sizes.push_back(tmp);
    }

    int maxs = 0;
    for (int i = 0; i < springs.size(); i++) {
        maxs = std::max((int) springs[i].size(), maxs);
        cout << springs[i] << " ";
        for (int j: group_sizes[i]) {
            cout << j << " ";
        }
        cout << endl;
    }
    cout << "max input size: " << maxs << endl;
    cout << endl;

    long long ans = 0;
    for (int i = 0; i < springs.size(); ++i) {
        vector<int> curr_sizes(group_sizes[i].size(), 0);
        long long ways = arrangements(springs[i], group_sizes[i], curr_sizes, 0, 0);
        cout << ways << endl;
        ans += ways;
    }
    cout << ans << endl;

}

void part2(ifstream &input) {
    string line;
    vector<string> springs;
    vector<vector<int>> group_sizes;
    while (getline(input, line)) {
        auto sp = filter(split(line, " "));

        springs.push_back(sp[0]);
        for (int i = 0; i < 4; ++i) {
            springs.back() += '?';
            springs.back() += sp[0];
        }

        vector<int> tmp = stoi(filter(split(sp[1], ",")));
        group_sizes.push_back(tmp);

        for (int i = 0; i < 4; ++i) {
            group_sizes.back().insert(group_sizes.back().end(), tmp.begin(), tmp.end());
        }


    }

    int maxs = 0;
    for (int i = 0; i < springs.size(); i++) {
        maxs = std::max((int) springs[i].size(), maxs);
        cout << springs[i] << " ";
        for (int j: group_sizes[i]) {
            cout << j << " ";
        }
        cout << endl;
    }
    cout << "max input size: " << maxs << endl;
    cout << endl;

    long long ans = 0;
//    vector<int> curr_sizes(group_sizes[1].size(), 0);
//    long long way = arrangements(springs[1], group_sizes[1], curr_sizes, 0, 0);
//    cout << "arr: " << way << endl;
    for (int i = 0; i < springs.size(); ++i) {
        vector<int> curr_sizes(group_sizes[i].size() + 1, 0);
        int maxi_group_length = *max_element(group_sizes[i].begin(), group_sizes[i].end());

        vector<vector<vector<long long>>> newdp(springs[i].size() + 1,
                                                vector<vector<long long>>(group_sizes[i].size() + 1,
                                                                          vector<long long>(maxi_group_length * 2,
                                                                                            -1)));
        dp = newdp;

        long long ways = arrangements_dp(springs[i], group_sizes[i], curr_sizes, 0, 0);
        cout << ways << endl;
        ans += ways;
    }
    cout << ans << endl;
}

int main() {
    ifstream input("/home/leartsura/Documents/projects/aoc23/input");
    if (input.fail()) {
        cout << "cant open input file" << endl;
        exit(0);
    }
    part1(input);
//    part2(input);

    input.close();
}
